### ISSUE TITLE
louvain graph community detection

### ISSUE THREAD

**Issue by ejblom** [graphology/graphology#429]
Hi all,

Quite pleased with the graphology library, especially with the sigma visualization. Thanks for developing it!
I tried to use the Louvain community detection algorithm but the resulting cliques are rather sparse, so not really interconnected.
Some of the members of a community only have a single edge to another member. 
If I compare the community sizes of other approaches, for instance in Gephi, I notice that other approaches create communities that are much larger and denser. 

I tried to play with the resolution setting, but that did not really provide consistent results. Are there some other tips and tricks that can be employed to get larger/denser clusters?

**Comment by Yomguithereal** [comment]
Hello @ejblom, without some visual examples nor more details I am afraid it will be complicated to help you much. With what kind of networks are you working?

> I tried to use the Louvain community detection algorithm but the resulting cliques are rather sparse, so not really interconnected.

By cliques you mean communities? The found communities should be internally denser than with their relationship to other communities.

> Some of the members of a community only have a single edge to another member.

This is not compulsorily a problem as the Louvain algorithm often produce this kind of relationships. Maybe your edges are weighted and some have more importance that you may think? Maybe you want to apply the unweighted algorithm instead (the weighted one being the default).

> If I compare the community sizes of other approaches, for instance in Gephi, I notice that other approaches create communities that are much larger and denser.

Other approaches of Louvain or of other community detection algorithms?

>I tried to play with the resolution setting, but that did not really provide consistent results. Are there some other tips and tricks that can be employed to get larger/denser clusters?

Note that graphology's resolution parameter follows the papers and is therefore reverse than what can be found on Gephi where they chose to align the direction of the parameter with intuition. Resolution can be quite finicky to tweak and resolution limit issues will sometimes mean that wanting larger or denser clusters amount to have a single community being the whole graph in specific cases.

**Comment by ejblom** [comment]
Hi @Yomguithereal thanks for the fast response, much appreciated. Sorry for the confusion, I mixed cliques and communities indeed. I don't have any great visuals to share yet but I'll try to be more precise in my wording :-).

The edges I use don't have any weight linked to them, although I am quite interested in the concept since I could foresee some future applications. You mention that the weighted algorithm is being used by default, could you indicate how to employ the unweighted version instead?

I was not aware of the difference between the Louvain implementation on Gephi and the one from Graphology. Also, after closer inspection of the sparse clusters, I think I could always apply some filtering to remove some nodes that have low-edge linkage to the other members of the community. I don't think that should be hard to implement.

Now that I think of it, perhaps the Louvain algorithm by default tries to include as many nodes as possible, even if it means that they only link with a single or two edges to the community. As you might expect from my questions, I don't really mind losing nodes if they don't have enough links to be part of the cluster.

**Comment by Yomguithereal** [comment]
> You mention that the weighted algorithm is being used by default, could you indicate how to employ the unweighted version instead?

Sure, just pass the `getEdgeWeight` setting to `null` like so:

```js
louvain(graph, {getEdgeWeight: null});
```

> I was not aware of the difference between the Louvain implementation on Gephi and the one from Graphology.

Apart from some parameters differences, they should be quite similar though. But since the Louvain algorithm is not deterministic and gives you a different solution each time, it might be difficult to compare partitions from Gephi and from graphology. Though, you seem to witness some macro differences that I would love to visually see (if they are not caused by weight) because it might point to some differences wrt optimizations used by both implementations. Louvain implementations are a fascinating topic and we are currently seeing that some might make things more or less deterministic in a tradeoff for performance or convergence.

> Now that I think of it, perhaps the Louvain algorithm by default tries to include as many nodes as possible, even if it means that they only link with a single or two edges to the community. As you might expect from my questions, I don't really mind losing nodes if they don't have enough links to be part of the cluster.

The algorithm will merge as many communities it can if it yields an increase of the graph's modularity. So if a node is linked by one or two edges to nodes of a large community it will be merged with it if this same node is not linked to another community with more edges or a larger total weight. Gephi should not work differently. If you do observe something off, one of both implementations might be faulty in some way.

**Comment by ejblom** [comment]
Thanks for the insights @Yomguithereal , I emailed you some more details btw. I implemented a filter that removed matches with low coverage to a community and this seems to improve things quite a bit.

**Comment by Yomguithereal** [comment]
I have seen your email @ejblom and I see your point. What I don't see is whether your connectivity problem is a byproduct of the algorithm or of my implementation. One thing that could help me would be to see comparisons of results from Gephi and from my implementation to see if maybe I did something wrong in the code at some point.

(I agree that Leiden should produce better connected solutions as it was basically designed to do so. I want to implement it in graphology but I lack proper time to do so currently).

**Comment by ejblom** [comment]
> I have seen your email @ejblom and I see your point. What I don't see is whether your connectivity problem is a byproduct of the algorithm or of my implementation. One thing that could help me would be to see comparisons of results from Gephi and from my implementation to see if maybe I did something wrong in the code at some point.
> 
> (I agree that Leiden should produce better connected solutions as it was basically designed to do so. I want to implement it in graphology but I lack proper time to do so currently).

You are right and I might not have been very clear on this. The Leiden algorithm is also implemented in Gephi and I might have seen examples of the well-connected clusters from this approach. So in that respect, the comparison isn't really valid since I was comparing apples and pears. Still, I was about to give up on the Louvain clustering but after this discussion, I had the idea to implement the additional filtering, which made it quite usable.

**Comment by Yomguithereal** [comment]
Can you elaborate about this filtering? What do you actually do here?

**Comment by ejblom** [comment]
> Can you elaborate about this filtering? What do you actually do here?

Oh really simple. So let's imagine a community has 5 members. I set a threshold that a member needs to link to at least 4 members, so a minimum linkage of 0.8. I simply check the links for each node with the remaining nodes. I just realize, there is no link with the node itself, so I should fix that. But that's basically the procedure. I bet it could be done by checking the edges between the node and the others, I just used the data that was used to generate the edges/nodes.

**Comment by Yomguithereal** [comment]
That should leave a lot of nodes outside of any community, no? I am not saying this is a bad thing but this is the first time I see someone with this requirement.

**Comment by ejblom** [comment]
> That should leave a lot of nodes outside of any community, no? I am not saying this is a bad thing but this is the first time I see someone with this requirement.

Yes, I guess it's all linked to the question you wish to answer. For my applications, I am not that interested in finding communities of let's say 20 members where some of its members are only linked with a single edge. I even have developed another application that required communities that are even more dense, so almost fully connected.

**Comment by abundis-rmn2** [comment]
Just in case it may be useful i use this code to add color to each node for specific community, may be not that elegant but works.

Define color for each louvain community
```
let com1= "#4ECDC4";
let com2= "#C7F464";
let com3= "#FF6B6B";
let com4= "#C44D58";
let com5= "#53777A";
let com6= "#78ec97";
let com0= "#556270";
```

Calculate Louvain Communities with graphology Library
and assign to the graph (each node)
```
const louvain =  new graphologyLibrary.communitiesLouvain(graph);
graphologyLibrary.communitiesLouvain.assign(graph);
```
For each node in `const louvain` set node color
```
Object.entries(louvain).forEach((entry) => {
    console.log("colorear");
    let node = entry[0]
    let community = entry[1]
    
    if (community == 0 ) {
        graph.setNodeAttribute(entry[0], 'color', com0);
    } else if (community == 1) {
        graph.setNodeAttribute(entry[0], 'color', com1);
    } else if (community == 2) {
        graph.setNodeAttribute(entry[0], 'color', com2);
    } else if (community == 3) {
        graph.setNodeAttribute(entry[0], 'color', com3);
    } else if (community == 4) {
        graph.setNodeAttribute(entry[0], 'color', com4);
    } else if (community == 5) {
        graph.setNodeAttribute(entry[0], 'color', com5);
    } else if (community == 6) {
        graph.setNodeAttribute(entry[0], 'color', com6);
    } else {
        graph.setNodeAttribute(entry[0], 'color', '#cccccc');
    }
});
```

![image](https://user-images.githubusercontent.com/67477259/198707591-0c38ec27-66e5-4c67-a4ec-d9aad1fd0b26.png)
[Working link](https://data.abundis.com.mx/vista/sigma.php?id=41)

By the way, i filter missing, or not relevant nodes dropping them if they have a small degree, you can change it for in or out degree also.
```
graph.nodes().forEach((node, i) => {
    if (graph.degree(node) <= 10){
        drop_cont++; 
        graph.dropNode(node);
        //console.log(drop_cont);
    } else {
    graph.setNodeAttribute(node, 'size', graph.inDegree(node) / 10) + 1;
    }
});
```

**Comment by Yomguithereal** [comment]
Thanks for the example @abundis-rmn2. I personally tend to use https://medialab.github.io/iwanthue/ to chose colors automagically based on the number of resulting communities. It has a npm package here: https://www.npmjs.com/package/iwanthue

Just a warning though, here:

```js
const louvain =  new graphologyLibrary.communitiesLouvain(graph);
graphologyLibrary.communitiesLouvain.assign(graph);
```

You compute communities, then you run the algorithm a second time to assign a new, potentially different set of communities, to the graph. By default the Louvain algorithm is not deterministic and usually returns slightly different (or widely different in some bad edge cases) results each time it runs. If you want to avoid that you can use `randomWalk` to false or provide your own seeded rng function as described here: https://graphology.github.io/standard-library/communities-louvain

But the simplest is probably to assign the result of running the algorithm once yourself in the subsequent loop you use to set node colors.
